File: SegmentManager.Textures.cpp
// Segment.cpp

#include "mine.h"
#include "dle-xp.h"

extern short nDbgSeg, nDbgSide;
extern int nDbgVertex;

// -----------------------------------------------------------------------------

void CSegmentManager::Textures (CSideKey key, short& nBaseTex, short& nOvlTex)
{
current->Get (key);
Side (key)->GetTextures (nBaseTex, nOvlTex);
}

// -----------------------------------------------------------------------------

bool CSegmentManager::SetTextures (CSideKey key, int nBaseTex, int nOvlTex)
{
	bool bChange = false;

undoManager.Begin (__FUNCTION__, udSegments); 
current->Get (key); 
CSide *sideP = Side (key); 
bChange = sideP->SetTextures (nBaseTex, nOvlTex);
if (!bChange) {
	undoManager.End (__FUNCTION__);
	return false;
	}
if ((lightManager.IsLight (sideP->BaseTex ()) == -1) && (lightManager.IsLight (sideP->OvlTex (0)) == -1))
	lightManager.DeleteVariableLight (key); 
if (!wallManager.ClipFromTexture (key))
	wallManager.CheckForDoor (key); 
undoManager.End (__FUNCTION__); 
sprintf_s (message, sizeof (message), "side has textures %d, %d", sideP->BaseTex () & 0x1fff, sideP->OvlTex (0)); 
INFOMSG (message); 
return true;
}

//------------------------------------------------------------------------------

void CSegmentManager::AlignTextures (short nSegment, short nSide, int bUse1st, int bUse2nd, int bIgnorePlane, bool bStart, bool bTagged)
{
if (bStart) {
	segmentManager.UnTagAll (ALIGNED_MASK);
	CSegment* segP = segmentManager.Segment (0);
	for (int i = 0; i < segmentManager.Count (); i++, segP++)
		for (short j = 0; j < 6; j++)
			// if we're aligning marked sides, consider untagged already aligned
			if ((segP->Side (j)->Shape () > SIDE_SHAPE_TRIANGLE) || (bTagged && !segmentManager.IsTagged (CSideKey (i, j))))
				segP->Tag (j, ALIGNED_MASK);
	}
// mark current side as aligned
segmentManager.Segment (nSegment)->Tag (nSide, ALIGNED_MASK);
// call recursive function which aligns one at a time
AlignChildTextures (nSegment, nSide, bUse1st, bUse2nd, bIgnorePlane);
}

//------------------------------------------------------------------------------

void CSegmentManager::AlignChildTextures (short nSegment, short nSide, int bUse1st, int bUse2nd, int bIgnorePlane)
{
	CSegment*		segP = segmentManager.Segment (nSegment);
	CSide*			sideP = segP->Side (nSide);
	CTagByTextures tagger (bUse1st ? sideP->BaseTex () : -1, bUse2nd ? sideP->OvlTex () : -1, bIgnorePlane);
	int				nSides;

if (!tagger.Setup (segmentManager.VisibleSideCount (), ALIGNED_MASK))
	return;
nSides = tagger.Run ();

for (int i = 0; i < nSides; i++) 
	segmentManager.AlignSideTextures (tagger.ParentSegment (i), tagger.ParentSide (i), tagger.ChildSegment (i), tagger.ChildSide (i), bUse1st, bUse2nd);

segmentManager.UnTagAll (TAGGED_MASK | ALIGNED_MASK);
}

// ------------------------------------------------------------------------ 

int CSegmentManager::AlignSideTextures (short nStartSeg, short nStartSide, short nChildSeg, short nChildSide, int bAlign1st, int bAlign2nd)
{
#ifdef _DEBUG
if ((nStartSeg < 0) || (nStartSeg >= Count ()))
	return -1;
if ((nStartSide < 0) || (nStartSide >= 6))
	return -1;
if ((nChildSeg < 0) || (nChildSeg >= Count ()))
	return -1;
if ((nChildSide < 0) || (nChildSide >= 6))
	return -1;

if ((nStartSeg == nDbgSeg) && ((nDbgSide < 0) || (nStartSide == nDbgSide)))
	nDbgSeg = nDbgSeg;
#endif

CSegment*	segP = Segment (nStartSeg); 
CSide*		sideP = segP->m_sides + nStartSide; 

if (sideP->Shape () > SIDE_SHAPE_TRIANGLE)
	return -1;

CSegment*	childSegP = Segment (nChildSeg); 
CSide*		childSideP = childSegP->m_sides + nChildSide; 
short			nEdgeCount = sideP->VertexCount ();
short			nChildEdgeCount = childSideP->VertexCount ();

short			i; 
short			nEdge, nChildEdge; 

// ignore children with no texture
if (!IsWall (CSideKey (nChildSeg, nChildSide)))
	return -1;

// find matching lines first
for (nEdge = 0; nEdge < nEdgeCount; nEdge++) {
	// find vert numbers for the line's two end points
	ushort vert0 = segP->VertexId (nStartSide, nEdge); 
	ushort vert1 = segP->VertexId (nStartSide, nEdge + 1); // the index (here: nEdge + 1) gets automatically wrapped by this function

	for (nChildEdge = 0; nChildEdge < nChildEdgeCount; nChildEdge++) {
		// find vert numbers for the line's two end points
		ushort nChildVert0 = childSegP->VertexId (nChildSide, nChildEdge);
		ushort nChildVert1 = childSegP->VertexId (nChildSide, nChildEdge + 1);
		// if points of child's line == corresponding points of parent
		if (((nChildVert0 == vert1) && (nChildVert1 == vert0)) || ((nChildVert0 == vert0) && (nChildVert1 == vert1)))
			break;
		}
	if (nChildEdge != nChildEdgeCount)
		break;
	}
if ((nEdge == nEdgeCount) || (nChildEdge == nChildEdgeCount))
	return -1;

// now we know the child's side & line which touches the parent
// child:  nChildSeg, nChildSide, nChildEdge, nChildPoint0, nChildPoint1
// parent: nStartSeg, nStartSide, nEdge, point0, point1
// we can start undoManager now - since we know we are making changes now and won't abort inside this block
undoManager.Begin (__FUNCTION__, udSegments);

double angle;

if (bAlign1st) {
	CUVL uv;
	// now translate all the childs (u, v) coords so the reference point is at zero
	uv = childSideP->m_info.uvls [(nChildEdge + 1) % nChildEdgeCount]; 
	for (i = 0; i < nChildEdgeCount; i++)
		childSideP->m_info.uvls [i] -= uv; 
	// find difference between parent point0 and child point1
	uv = childSideP->m_info.uvls [(nChildEdge + 1) % nChildEdgeCount] - sideP->m_info.uvls [nEdge]; 
#if 1
	// "round" to the -20 to 20 neighborhood
	uv.u = fmod (uv.u, 1);
	uv.v = fmod (uv.v, 1);
#endif
	// find the angle formed by the two lines
	double sAngle = atan3 (sideP->m_info.uvls [(nEdge + 1) % nEdgeCount].v - sideP->m_info.uvls [nEdge].v, 
								  sideP->m_info.uvls [(nEdge + 1) % nEdgeCount].u - sideP->m_info.uvls [nEdge].u); 
	double cAngle = atan3 (childSideP->m_info.uvls [nChildEdge].v - childSideP->m_info.uvls [(nChildEdge + 1) % nChildEdgeCount].v, 
					 			  childSideP->m_info.uvls [nChildEdge].u - childSideP->m_info.uvls [(nChildEdge + 1) % nChildEdgeCount].u); 
	// now rotate childs (u, v) coords around child_point1 (cAngle - sAngle)
	angle = cAngle - sAngle;
	for (i = 0; i < nChildEdgeCount; i++) 
		childSideP->m_info.uvls [i].Rotate (angle); 
	// now translate all the childs (u, v) coords to parent point0
	for (i = 0; i < nChildEdgeCount; i++)
		childSideP->m_info.uvls [i] -= uv; 
	}
if (bAlign2nd && sideP->OvlTex (0) && childSideP->OvlTex (0)) {
	int r, h = sideP->OvlAlignment ();
	if (h == 3)
		r = 1;
	else if (h == 2)
		r = 2;
	else if (h == 1)
		r = 3;
	else
		r = 0;
	if (angle < 0.0)
		angle += 2 * PI;
	h = (int) Round (Degrees (angle) / 90.0); 
	h = (h + r) % nEdgeCount;
	childSideP->m_info.nOvlTex &= TEXTURE_MASK;
	if (h == 1)
		childSideP->m_info.nOvlTex |= 0xC000;
	else if (h == 2)
		childSideP->m_info.nOvlTex |= 0x8000;
	else if (h == 3)
		childSideP->m_info.nOvlTex |= 0x4000;
	}

undoManager.End (__FUNCTION__);
return nChildSide; 
}

// -----------------------------------------------------------------------------

void CSegmentManager::UpdateTexCoords (ushort nVertexId, bool bMove, short nIgnoreSegment, short nIgnoreSide)
{
CSegment* segP = Segment (0);
for (short nSegment = 0; 0 <= (nSegment = FindByVertex (nVertexId, nSegment)); nSegment++, segP++) {
	if (nIgnoreSegment >= 0) {
		if ((nIgnoreSide < 0) && (nSegment == nIgnoreSegment))
			continue;
		if ((nIgnoreSegment == 0x7FFF) && segP->IsTagged ())
			continue;
		}
	segP->UpdateTexCoords (nVertexId, bMove, (nSegment == nIgnoreSegment) ? nIgnoreSide : -1);
	}
}

// ------------------------------------------------------------------------
// ------------------------------------------------------------------------
// ------------------------------------------------------------------------
//eof segmentmanager.textures.cpp

File: DLE.NET/Segment.Manager.Textures.cs
using System;

namespace DLE.NET
{
    public partial class SegmentManager
    {
        // ------------------------------------------------------------------------

        public void Textures (SideKey key, out short nBaseTex, out short nOvlTex)
        {
        DLE.Current.Get (key);
        Side (key).GetTextures (out nBaseTex, out nOvlTex);
        }

        // ------------------------------------------------------------------------

        public bool SetTextures (SideKey key, ushort nBaseTex, ushort nOvlTex)
        {
	        bool bChange = false;

        DLE.Backup.Begin (UndoData.Flags.udSegments); 
        DLE.Current.Get (key); 
        Side side = Side (key);
        bChange = side.SetTextures ((ushort)nBaseTex, (ushort)nOvlTex);
        if (!bChange) {
	        DLE.Backup.End ();
	        return false;
	        }
        if ((DLE.Lights.IsLight (side.m_nBaseTex) == -1) && (DLE.Lights.IsLight (side.m_nOvlTex & 0x3fff) == -1))
	        DLE.Lights.DeleteVariableLight (key); 
        if (!DLE.Walls.ClipFromTexture (key))
	        DLE.Walls.CheckForDoor (key); 
        DLE.Backup.End (); 
        //sprintf_s (message, sizeof (message), "side has textures %d, %d", side.m_nBaseTex & 0x1fff, side.m_nOvlTex & 0x1fff); 
        //DLE.InfoMsg (message); 
        return true;
        }

        // ------------------------------------------------------------------------

        static short [,] sideChildTable = new short [6, 4] {
		        {4, 3, 5, 1}, //{5, 1, 4, 3}, 
		        {2, 4, 0, 5}, //{5, 0, 4, 2}, 
		        {5, 3, 4, 1}, //{5, 3, 4, 1}, 
		        {0, 4, 2, 5}, //{5, 0, 4, 2}, 
		        {2, 3, 0, 1}, //{2, 3, 0, 1}, 
		        {0, 3, 2, 1} //{2, 3, 0, 1}
		        };

        public int AlignTextures (short nStartSeg, short nStartSide, short nOnlyChildSeg, bool bAlign1st, bool bAlign2nd, sbyte bAlignedSides = 0)
        {
	        Segment	    seg = Segments [nStartSeg]; 
	        Segment	    childSeg; 
	        Side		side = seg.m_sides [nStartSide]; 
	        Side		childSide; 

	        int			return_code = -1; 
	        short		i; 
	        short		nSide, nChildSeg, nLine; 
	        ushort		point0, point1, vert0, vert1; 
	        short		nChildSide, nChildLine; 
	        ushort		nChildPoint0, nChildPoint1, nChildVert0, nChildVert1; 
	        short		u0, v0; 
	        double		sAngle, cAngle, angle, length; 

        DLE.Backup.Begin (UndoData.Flags.udSegments);
        for (nLine = 0; nLine < 4; nLine++) {
	        // find vert numbers for the line's two end points
	        point0 = GameTables.lineVertTable [GameTables.sideLineTable [nStartSide,nLine],0]; 
	        point1 = GameTables.lineVertTable [GameTables.sideLineTable [nStartSide,nLine],1]; 
	        vert0 = seg.m_verts [point0]; 
	        vert1 = seg.m_verts [point1]; 
	        // check child for this line
	        if (nStartSeg == nOnlyChildSeg) {
		        nSide = nStartSide;
		        nChildSeg = nStartSeg;
		        }
	        else {
		        nSide = sideChildTable [nStartSide,nLine]; 
		        nChildSeg = seg.GetChild (nSide); 
		        }
	        if ((nChildSeg < 0) || ((nOnlyChildSeg != -1) && (nChildSeg != nOnlyChildSeg)))
		        continue;
	        childSeg = Segments [nChildSeg]; 
	        // figure out which side of child shares two points w/ nStartSide
	        for (nChildSide = 0; nChildSide < 6; nChildSide++) {
		        if ((nStartSeg == nOnlyChildSeg) && (nChildSide == nStartSide))
			        continue;
		        if ((bAlignedSides & (sbyte) (1 << nChildSide)) != 0)
			        continue;
		        // ignore children of different textures (or no texture)
		        if (!IsWall (new SideKey (nChildSeg, nChildSide)))
			        continue;
		        if (childSeg.m_sides [nChildSide].m_nBaseTex != side.m_nBaseTex)
			        continue;
		        for (nChildLine = 0; nChildLine < 4; nChildLine++) {
			        // find vert numbers for the line's two end points
                    nChildPoint0 = GameTables.lineVertTable [GameTables.sideLineTable [nChildSide,nChildLine],0];
                    nChildPoint1 = GameTables.lineVertTable [GameTables.sideLineTable [nChildSide,nChildLine],1]; 
			        nChildVert0  = childSeg.m_verts [nChildPoint0]; 
			        nChildVert1  = childSeg.m_verts [nChildPoint1]; 
			        // if points of child's line== corresponding points of parent
			        if (!((nChildVert0 == vert1 && nChildVert1 == vert0) ||
					        (nChildVert0 == vert0 && nChildVert1 == vert1)))
				        continue;
			        // now we know the child's side & line which touches the parent
			        // child:  nChildSeg, nChildSide, nChildLine, nChildPoint0, nChildPoint1
			        // parent: nStartSeg, nStartSide, nLine, point0, point1
			        childSide = childSeg.m_sides [nChildSide]; 
			        if (bAlign1st) {
				        // now translate all the childs (u, v) coords so child_point1 is at zero
				        u0 = childSide.m_uvls [(nChildLine + 1) % 4].u; 
				        v0 = childSide.m_uvls [(nChildLine + 1) % 4].v; 
				        for (i = 0; i < 4; i++) {
					        childSide.m_uvls [i].u -= u0; 
					        childSide.m_uvls [i].v -= v0; 
					        }
				        // find difference between parent point0 and child point1
                        u0 = (short) (childSide.m_uvls [(nChildLine + 1) % 4].u - side.m_uvls [nLine].u);
                        v0 = (short) (childSide.m_uvls [(nChildLine + 1) % 4].v - side.m_uvls [nLine].v); 
				        // find the angle formed by the two lines
				        sAngle = Math.Atan2((double) (side.m_uvls [(nLine + 1) % 4].v - side.m_uvls [nLine].v), 
									        (double) (side.m_uvls [(nLine + 1) % 4].u - side.m_uvls [nLine].u)); 
				        cAngle = Math.Atan2((double) (childSide.m_uvls [nChildLine].v - childSide.m_uvls [(nChildLine + 1) % 4].v), 
									        (double) (childSide.m_uvls [nChildLine].u - childSide.m_uvls [(nChildLine + 1) % 4].u)); 
				        // now rotate childs (u, v) coords around child_point1 (cangle - sangle)
				        for (i = 0; i < 4; i++) {
					        angle = Math.Atan2((double) childSide.m_uvls [i].v, (double) childSide.m_uvls [i].u); 
					        length = Math.Sqrt((double) childSide.m_uvls [i].u * (double) childSide.m_uvls [i].u +
									          (double) childSide.m_uvls [i].v * (double) childSide.m_uvls [i].v); 
					        angle -= (cAngle - sAngle); 
					        childSide.m_uvls [i].u = (short)(length * Math.Cos (angle)); 
					        childSide.m_uvls [i].v = (short)(length * Math.Sin (angle)); 
					        }
				        // now translate all the childs (u, v) coords to parent point0
				        for (i = 0; i < 4; i++) {
					        childSide.m_uvls [i].u -= u0; 
					        childSide.m_uvls [i].v -= v0; 
					        }
				        // modulo points by 0x800 (== 64 pixels)
				        u0 = (short) (childSide.m_uvls [0].u / 0x800); 
				        v0 = (short) (childSide.m_uvls [0].v / 0x800); 
				        for (i = 0; i < 4; i++) {
					        childSide.m_uvls [i].u -= (short) (u0 * 0x800); 
					        childSide.m_uvls [i].v -= (short) (v0 * 0x800); 
					        }
				        if (nOnlyChildSeg != -1)
					        return_code = nChildSide; 
				        }
			        if (bAlign2nd && (side.m_nOvlTex != 0) && (childSide.m_nOvlTex != 0)) {
                        int r;
                        int h = side.m_nOvlTex & 0xC000;
                        if (h == 0xC000)
						    r = 1;
					    else if (h == 0x8000)
						    r = 2;
					    else if (h == 0x4000)
						    r = 3;
						else
                            r = 0;
                        angle = Math.Atan2 ((double)childSide.m_uvls [0].v, (double)childSide.m_uvls [0].u);
                        h = (int)(MathExtensions.Degrees (Math.Abs (angle)) / 90 + 0.5); 
        //				h +=(nChildLine + nLine + 2) % 4; //(nChildLine > nLine) ? nChildLine - nLine : nLine - nChildLine;
				        h = (h + r) % 4;
                        unchecked
                        {
                            childSide.m_nOvlTex &= (ushort)~0xC000;
                        }
				        if (h == 1)
						    childSide.m_nOvlTex |= 0xC000;
				        else if (h == 2)
						    childSide.m_nOvlTex |= 0x8000;
				        else if (h == 3)
						    childSide.m_nOvlTex |= 0x4000;
				        }
			        break;
			        }
		        }
	        }
        DLE.Backup.End ();
        return return_code; 
        }

        // ------------------------------------------------------------------------
    }
}